src/HTTP.jl

```diff
@@ -1,83 +1,586 @@
+
+ - `require_ssl_verification = false`, pass `MBEDTLS_SSL_VERIFY_REQUIRED` to
+   the mbed TLS library.
+   ["... peer must present a valid certificate, handshake is aborted if
+     verification failed."](https://tls.mbed.org/api/ssl_8h.html#a5695285c9dbfefec295012b566290f37)
+ - sslconfig = SSLConfig(require_ssl_verification)`
+
+
+Basic Authenticaiton options
+
+ - basicauthorization=false, add `Authorization: Basic` header using credentials
+   from url userinfo.
+
+
+AWS Authenticaiton options
+
+ - `awsauthorization = false`, enable AWS4 Authentication.
+ - `aws_service = split(uri.host, ".")[1]`
+ - `aws_region = split(uri.host, ".")[2]`
+ - `aws_access_key_id = ENV["AWS_ACCESS_KEY_ID"]`
+ - `aws_secret_access_key = ENV["AWS_SECRET_ACCESS_KEY"]`
+ - `aws_session_token = get(ENV, "AWS_SESSION_TOKEN", "")`
+ - `body_sha256 = digest(MD_SHA256, body)`,
+ - `body_md5 = digest(MD_MD5, body)`,
+
+
+Cookie options
+
+ - `cookies = false`, enable cookies.
+ - `cookiejar::Dict{String, Set{Cookie}}=default_cookiejar`
+
+
+Cananoincalization options
+
+ - `canonicalizeheaders = false`, rewrite request and response headers in
+   Canonical-Camel-Dash-Format.
+
+
+## Request Body Examples
+
+String body:
+```julia
+request("POST", "http://httpbin.org/post", [], "post body data")
+```
+
+Stream body from file:
+```julia
+io = open("post_data.txt", "r")
+request("POST", "http://httpbin.org/post", [], io)
+```
+
+Generator body:
+```julia
+chunks = ("chunk\$i" for i in 1:1000)
+request("POST", "http://httpbin.org/post", [], chunks)
+```
+
+Collection body:
+```julia
+chunks = [preamble_chunk, data_chunk, checksum(data_chunk)]
+request("POST", "http://httpbin.org/post", [], chunks)
+```
+
+`open() do io` body:
+```julia
+HTTP.open("POST", "http://httpbin.org/post") do io
+    write(io, preamble_chunk)
+    write(io, data_chunk)
+    write(io, checksum(data_chunk))
 end
+```
 
-macro uninit(expr)
-    if !isdefined(Base, :uninitialized)
-        splice!(expr.args, 2)
-    end
-    return esc(expr)
+
+## Response Body Examples
+
+String body:
+```julia
+r = request("GET", "http://httpbin.org/get")
+println(String(r.body))
+```
+
+Stream body to file:
+```julia
+io = open("get_data.txt", "w")
+r = request("GET", "http://httpbin.org/get", response_stream=io)
+close(io)
+println(read("get_data.txt"))
+```
+
+Stream body through buffer:
+```julia
+io = BufferStream()
+@async while !eof(io)
+    bytes = readavailable(io))
+    println("GET data: \$bytes")
 end
+r = request("GET", "http://httpbin.org/get", response_stream=io)
+close(io)
+```
 
-if !isdefined(Base, :pairs)
-    pairs(x) = x
+Stream body through `open() do io`:
+```julia
+r = HTTP.open("GET", "http://httpbin.org/stream/10") do io
+   while !eof(io)
+       println(String(readavailable(io)))
+   end
 end
 
-if !isdefined(Base, :Nothing)
-    const Nothing = Void
-    const Cvoid = Void
+HTTP.open("GET", "https://tinyurl.com/bach-cello-suite-1-ogg") do http
+    n = 0
+    r = startread(http)
```

Comment:
is `startread` exported? Better prefix with `HTTP.` anyway so users are confused about what this is.

Reply:
No, it isn't. I've added `using HTTP.IOExtras` to the doc string.
I would like to see something like `startread`, `closeread`, `startewrite`, `closewrite` in Base eventually. There are many IO protocols that have separate read/write open/close signalling  semantics, so it would be good to have a common API for this. e.g. [shutdown(fd, SHUT_WR)](http://man7.org/linux/man-pages/man2/shutdown.2.html)
See comments here https://github.com/JuliaLang/julia/issues/24526

Reply:
499b752128e9c99427ab6408793ecea541e97b0a

Reply:
Did you try it? did it make music?